src/collocation.jl

```diff
@@ -35,13 +35,83 @@ end
     end
 end
 
-function Φ(cache::MIRKCache, y, u, p = cache.p)
+@views function Φ!(residual, fᵢ_cache, k_discrete, f!, TU::RKTableau{false}, y, u, p,
+                   mesh, mesh_dt, stage::Int)
+    @unpack c, a, b = TU
+    tmp1 = get_tmp(fᵢ_cache, u)
+    K = get_tmp(k_discrete[1], u) # Not optimal
+    T = eltype(u)
+    ctr = 1
+    for i in eachindex(k_discrete)
+        h = mesh_dt[i]
+        yᵢ = get_tmp(y[ctr], u)
+        yᵢ₊₁ = get_tmp(y[ctr + stage + 1], u)
+
+        # Load interpolation residual
+        for j in 1:stage
+            K[:, j] = get_tmp(y[ctr + j], u)
+        end
+
+        # Update interpolation residual
+        for r in 1:stage
+            @. tmp1 = yᵢ
+            __maybe_matmul!(tmp1, K[:, 1:stage], a[r, 1:stage], h, T(1))
+            f!(residual[ctr + r], tmp1, p, mesh[i] + c[r] * h)
+            residual[ctr + r] .-= K[:, r]
+        end
+
+        # Update mesh point residual
+        residᵢ = residual[ctr]
+        @. residᵢ = yᵢ₊₁ - yᵢ
+        __maybe_matmul!(residᵢ, K[:, 1:stage], b[1:stage], -h, T(1))
+        ctr += stage + 1
+    end
+end
+
+function FIRK_nlsolve(K, f!, a, c, yᵢ, h, mesh_i, stage, p)
+    res = copy(K)
+    T = eltype(K)
+    tmp1 = similar(K, size(K, 1))
+
+    for r in 1:stage
+        @. tmp1 = yᵢ
+        __maybe_matmul!(tmp1, K[:, 1:stage], a[r, 1:stage], h, T(1))
+        f!(@view(res[:, r]), tmp1, p, mesh_i + c[r] * h)
+        res[:, r] .-= K[:, r]
+    end
+    return res
+end
+
+@views function Φ!(residual, fᵢ_cache, k_discrete, f!, TU::RKTableau{true}, y, u, p,
+                   mesh, mesh_dt, stage::Int)
+    @unpack c, a, b = TU
+    T = eltype(u)
+    K = get_tmp(k_discrete[1], u)
+
+    for i in eachindex(k_discrete)
+        residᵢ = residual[i]
+        h = mesh_dt[i]
+
+        yᵢ = get_tmp(y[i], u)
+        yᵢ₊₁ = get_tmp(y[i + 1], u)
+        y_i = eltype(yᵢ) == Float64 ? yᵢ : [y.value for y in yᵢ] 
+        prob = NonlinearProblem((K, p) -> FIRK_nlsolve(K, f!, a, c, y_i, h, mesh[i], stage, p), fill(1.0, size(K)), p);
```

Comment:
This is the part that is giving me headaches conceptually right now. How can I avoid this closure-fest?

Reply:
(which is due to iterating over y, h and mesh)